### PR TITLE
Support for Gateway sanitization of image tags to digest

### PR DESCRIPTION
**Pull request by andrew-leung**
### Description of the Change
This PR provides basic support for gateway sanitization of image tags to digests.

Apologies for the large PR. Some of the key takeaways:
- All JobDescriptor tags are resolved to digest at the gateway. This results in the Job defining both a tag and a digest value.
- The PR defaults to **disabling** the tag to digest resolution by default because there are Titus Agent side changes required to support this properly when Jobs request Metatron credentials.
- When enabled, the image sanitizer currently **fails open**. The only case where it does not fail open is when an image (checked either via tag or digest) does not exist. In this case, a non-existent image error is returned.
- The tag to digest sanitization is the only current "off box" sanitization that happens. This step happens prior to validation so that validation consumes the final, sanitized job descriptor. Sanitizers are setup to run in serial while validators are setup to run in parallel. To more clearly reflect this, ParallelValidator was renamed AggregatingValidator.

### CODE REVIEW

**Comment by gabrielhartmann** [comment]
@andrew-leung: Net +1,600 lines...

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+@ProxyConfiguration(types = {ProxyType.Logging, ProxyType.Spectator})
+public class DefaultDockerRegistryClient implements RegistryClient {
+
+    private static final Logger logger = LoggerFactory.getLogger(DefaultDockerRegistryClient.class);
+
+    private static final String acceptHeader = "Accept";
+    private static final String dockerManifestType = "application/vnd.docker.distribution.manifest.v2+json";
+    private static final String dockerDigestHeaderKey = "Docker-Content-Digest";
+
+    private static final Map<String, String> headers = Collections.unmodifiableMap(
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: Why not `Collections.singletonMap`?

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryClient.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+package com.netflix.titus.runtime.connector.registry;
+
+import rx.Single;
+
+/**
+ * Interface for a container image registry client.
+ */
+public interface RegistryClient {
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryClient.java]: `ContainerRegistryClient` or `ImageRegistryClient`?

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryConfigurationEndpointResolver.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+    private final TitusRegistryClientConfiguration titusRegistryClientConfiguration;
+
+    @Inject
+    public RegistryConfigurationEndpointResolver(TitusRegistryClientConfiguration configuration) {
+        this.titusRegistryClientConfiguration = configuration;
+    }
+
+    @Override
+    public Observable<URI> resolve() {
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryConfigurationEndpointResolver.java]: Is there a use case when this must be asynchronous operation? For Eureka it is local call as well.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidator.java [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+                        )));
+    }
+
+    @Override
+    public Mono<JobDescriptor> sanitize(JobDescriptor jobDescriptor) {
+        if (isDisabled()) {
+            return Mono.just(jobDescriptor);
+        }
+
+        return validateImage(jobDescriptor)
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidator.java]: `sanitizeImage`?

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+ * limitations under the License.
+ */
+
+package com.netflix.titus.runtime.endpoint.validator;
+
+import com.netflix.archaius.api.annotations.DefaultValue;
+import com.netflix.archaius.api.annotations.PropertyName;
+
+public interface JobImageValidatorConfiguration {
+    @PropertyName(name = "titus.validate.job.image.enabled")
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java]: Use class level `@Configuration(prefix=...)`, and remove the `@PropertyName` annotation. The name will be derived from the method name.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/TitusValidatorConfiguration.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.runtime.endpoint.validator;
+
+import com.netflix.archaius.api.annotations.DefaultValue;
+import com.netflix.archaius.api.annotations.PropertyName;
+
+public interface TitusValidatorConfiguration {
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/TitusValidatorConfiguration.java]: Same comment as for `JobImageValidatorConfiguration`.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/TitusValidatorModule.java [hunk 7]
```diff
@@ -31,10 +45,12 @@ TitusValidatorConfiguration getConfiguration(ConfigProxyFactory factory) {
...
-        return new ParallelValidator(
-                Duration.ofMillis(Long.parseLong(configuration.getTimeoutMs())),
+    public EntityValidator getJobValidator(TitusValidatorConfiguration configuration,
+                                           JobImageValidator jobImageValidator) {
+        return new AggregatingValidator(
+                configuration,
                 Collections.emptyList(), // No HARD validators
-                Arrays.asList(new PassJobValidator()));
+                Arrays.asList(new PassJobValidator()),
+                Arrays.asList(jobImageValidator));
```
Comment (tbak) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/TitusValidatorModule.java]: How will this work in the `EmbeddedTitusGateway`?

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+import reactor.core.publisher.Mono;
+
+/**
+ * This {@link EntityValidator} implementation ensures a job's appname matches a specific
+ * string. It is only used for testing purposes.
+ */
+public class AppNameSanitizer implements EntityValidator<JobDescriptor> {
+    public final static String desiredAppName = "desiredAppName";
+    private static final String ERR_FIELD = "fail-field";
+    private static final String ERR_DESCRIPTION = "The job does not have desired appname " + desiredAppName;
```
Comment (gabrielhartmann) [inline titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java]: Prefer `String.format("The job does not have desired appname '%s'", desiredAppName)` as a style of error message as it gives a clearer indication when an expected field is empty.

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import com.netflix.titus.api.jobmanager.model.job.JobDescriptor;
+import com.netflix.titus.common.model.validator.EntityValidator;
+import com.netflix.titus.common.model.validator.ValidationError;
+import reactor.core.publisher.Mono;
+
+/**
+ * This {@link EntityValidator} implementation ensures a job's appname matches a specific
+ * string. It is only used for testing purposes.
+ */
+public class AppNameSanitizer implements EntityValidator<JobDescriptor> {
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java]: Will do.

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/CapacityGroupSanitizer.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+import com.netflix.titus.api.jobmanager.model.job.JobDescriptor;
+import com.netflix.titus.common.model.validator.EntityValidator;
+import com.netflix.titus.common.model.validator.ValidationError;
+import reactor.core.publisher.Mono;
+
+/**
+ * This {@link EntityValidator} implementation ensures a job's capacity group matches a specific
+ * string. It is only used for testing purposes.
+ */
+public class CapacityGroupSanitizer implements EntityValidator<JobDescriptor> {
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/CapacityGroupSanitizer.java]: Will do.

File: titus-server-gateway/src/main/java/com/netflix/titus/gateway/startup/TitusGatewayModule.java [hunk 11]
```diff
@@ -78,6 +80,7 @@ protected void configure() {
 
         install(new ContainerEventBusModule());
 
+        install(new TitusRegistryModule());
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-server-gateway/src/main/java/com/netflix/titus/gateway/startup/TitusGatewayModule.java]: Yeah, it's a pretty generic name :-/ I'll update.

File: titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RetryableRestClient.java [hunk 12]
```diff
@@ -45,20 +51,27 @@
...
         this.timeoutMs = timeUnit.toMillis(timeout);
         this.retryDelayMs = timeUnit.toMillis(retryDelay);
+        this.noRetryStatuses = noRetryStatuses;
     }
 
     @Override
-    public <T> Observable<T> doGET(String relativeURI, TypeProvider<T> typeProvider) {
-        return runWithRetry(HttpMethod.GET, relativeURI, delegate.doGET(relativeURI, typeProvider));
+    public <T> Observable<T> doGET(String relativeURI, TypeProvider<T> type) {
+        return doGET(relativeURI, Collections.unmodifiableMap(new HashMap<>()), type);
```
Comment (corindwyer) [inline titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RetryableRestClient.java]: `Collections.unmodifiableMap(new HashMap<>())` -> `Collections.emptyMap()`- immutable empty map that is a constant such that you don't need to allocate.

File: titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RxNettyRestClient.java [hunk 13]
```diff
@@ -81,9 +82,14 @@ public RxNettyRestClient(EndpointResolver endpointResolver,
 
     @Override
     public <T> Observable<T> doGET(String relativeURI, TypeProvider<T> type) {
+        return doGET(relativeURI, Collections.unmodifiableMap(new HashMap<>()), type);
```
Comment (corindwyer) [inline titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RxNettyRestClient.java]: same as above

File: titus-server-gateway/src/test/java/com/netflix/titus/gateway/service/v3/internal/JobImageValidatorTest.java [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    @Before
+    public void setUp() {
+        when(configuration.getEnabled()).thenReturn(true);
+        when(registryClient.getImageDigest(anyString(), anyString())).thenReturn(Single.just(digest));
+        validator = new JobImageValidator(configuration, registryClient);
+    }
+
+    @Test
+    public void testJobWithTagResolution() {
+        when(registryClient.getImageDigest(anyString(), anyString())).thenReturn(Single.just(digest));
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-server-gateway/src/test/java/com/netflix/titus/gateway/service/v3/internal/JobImageValidatorTest.java]: Yeah, good point. I'll move all of the tests to using `StepVerifier`.

File: titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RetryableRestClientTest.java [hunk 15]
```diff
@@ -54,12 +62,12 @@
     @Before
     public void setUp() throws Exception {
         client = new RetryableRestClient(delegate,
-                RETRY_COUNT, REQ_TIMEOUT_MS, RETRY_DELAY_MS, TimeUnit.MILLISECONDS, clientMetric, testScheduler);
+                RETRY_COUNT, REQ_TIMEOUT_MS, RETRY_DELAY_MS, TimeUnit.MILLISECONDS, new HashSet<>(Collections.singletonList(noRetryStatus)), clientMetric, testScheduler);
```
Comment (corindwyer) [inline titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RetryableRestClientTest.java]: `new HashSet<>(Collections.singletonList(noRetryStatus))` -> `Collections.singleton()`

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/PassJobValidator.java [hunk 16]
```diff
@@ -10,7 +26,7 @@
 import reactor.core.publisher.Mono;
 
 /**
- * This {@link EntityValidator} implementation always causes validation to fail.  It is used as a default implementation which
+ * This {@link EntityValidator} implementation always causes validation to pass.  It is used as a default implementation which
  * should be overriden.
```
Comment (corindwyer) [inline titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/PassJobValidator.java]: overriden -> overridden

File: titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobSubmitAndControlNegativeTest.java [hunk 17]
```diff
@@ -174,6 +174,7 @@ public void testJobWithoutImage() throws Exception {
     @Test
     @Ignore("Until we support digests")
     public void testJobWithBothTagAndDigest() throws Exception {
+        // TODO(Andrew L): Enable this test.
```
Comment (corindwyer) [inline titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobSubmitAndControlNegativeTest.java]: Was this an accident or do you want to do this in another PR?

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+        this.titusRegistryClientConfiguration = configuration;
+
+        RxRestClients.Builder builder = RxRestClients.newBuilder("dockerRegistryClient", spectatorRegistry)
+                .endpointResolver(endpointResolver)
+                // We rely on the RxRestClient for timeout handling
+                .timeUnit(TimeUnit.MILLISECONDS)
+                .requestTimeout(titusRegistryClientConfiguration.getRegistryTimeoutMs())
+                .retryCount(titusRegistryClientConfiguration.getRegistryRetryCount())
+                .retryDelay(titusRegistryClientConfiguration.getRegistryRetryDelayMs())
+                .noRetryStatuses(new HashSet<>(Collections.singletonList(HttpResponseStatus.NOT_FOUND)));
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: `Collections.singleton`

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        if (titusRegistryClientConfiguration.isSecure()) {
+            try {
+                builder.sslEngineFactory(new NettySslContextEngineFactory(SslContextBuilder.forClient().build()));
+            } catch (SSLException e) {
+                logger.error("Unable configure Docker registry client SSL context: {}", e.getMessage());
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: add `, e` at the end to get full stack trace

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+                throw new RuntimeException("Error configuring SSL context", e);
+            }
+        }
+        this.restClient = builder.build();
+    }
+
+    /**
+     * Gets the Docker Version 2 Schema 2 Content Digest for the provide repository and reference. The
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: provide -> provided

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+     * reference may be an image tag or digest value. If the image does not exist or another error is
+     * encountered, an onError value is emitted.
+     */
+    public Single<String> getImageDigest(String repository, String reference) {
+        return registryRequestWithErrorHandling(restClient.doGET(buildRegistryUri(repository, reference), headers, TypeProviders.ofEmptyResponse()))
+                .map(RxHttpResponse::getHeaders)
+                .flatMap(stringListMap -> {
+                    if (stringListMap.containsKey(dockerDigestHeaderKey)) {
+                        return Observable.from(stringListMap.get("Docker-Content-Digest"));
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: use constant `dockerDigestHeaderKey` defined above

File: titus-server-gateway/src/test/java/com/netflix/titus/gateway/service/v3/internal/JobImageValidatorTest.java [hunk 22]
```diff
@@ -0,0 +1,219 @@
...
+
+        JobDescriptor<?> jobDescriptor = JobDescriptorGenerator.batchJobDescriptors()
+                .map(jd -> jd.but(d -> d.getContainer().toBuilder()
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-server-gateway/src/test/java/com/netflix/titus/gateway/service/v3/internal/JobImageValidatorTest.java]: There are two different kinds of job descriptors being tested, but that's probably small enough to still move to consts. Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java [hunk 23]
```diff
@@ -0,0 +1,129 @@
...
+     * Wraps an observable registry request with timeouts and error handling.
+     */
+    private <T> Observable<T> registryRequestWithErrorHandling(Observable<T> obs) {
+        return obs.timeout(titusRegistryClientConfiguration.getRegistryTimeoutMs(), TimeUnit.MILLISECONDS)
+                .onErrorResumeNext(throwable -> {
+                    if (throwable instanceof RxRestClientException) {
+                        if (((RxRestClientException)throwable).getStatusCode() == HttpResponseStatus.NOT_FOUND.code()) {
+                            return Observable.error(
+                                    new TitusRegistryException(TitusRegistryException.ErrorCode.IMAGE_NOT_FOUND,
+                                    "Image does not exist in registry"));
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: pass in image name so this message will be more useful

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/AggregatingValidator.java [hunk 24]
```diff
@@ -0,0 +1,132 @@
...
+    private final TitusValidatorConfiguration configuration;
+
+    private final Duration timeout;
+    private final Collection<EntityValidator<JobDescriptor>> hardValidators;
+    private final Collection<EntityValidator<JobDescriptor>> softValidators;
+
+    private final Collection<EntityValidator<JobDescriptor>> sanitizers;
+
+    /**
+     * A AggregatingValidator ensures that all its constituent validators or sanitizers complete within a specified
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/AggregatingValidator.java]: A -> An

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java [hunk 25]
```diff
@@ -0,0 +1,26 @@
...
+    @DefaultValue("false")
+    boolean getEnabled();
```
Comment (corindwyer) [inline titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java]: getEnabled -> isEnabled

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/jobmanager/client/GrpcJobManagementClient.java [hunk 26]
```diff
@@ -88,32 +107,37 @@ public GrpcJobManagementClient(JobManagementServiceGrpc.JobManagementServiceStub
...
-
-                    // Only emit an error on HARD validation errors
-                    errors = errors.stream().filter(error -> error.isHard()).collect(Collectors.toSet());
-
-                    if (!errors.isEmpty()) {
-                        return Observable.error(TitusServiceException.invalidJob(errors));
-                    } else {
-                        return requestObservable;
-                    }
+        Observable<com.netflix.titus.api.jobmanager.model.job.JobDescriptor> sanitizedCoreJobDescriptorObs =
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/jobmanager/client/GrpcJobManagementClient.java]: I moved the external validators/sanitzers to the Gateway specific class. Though Federation is not actually using the common Grpc client.

File: titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RxNettyRestClientTest.java [hunk 27]
```diff
@@ -70,7 +71,7 @@ public void testGET() throws Exception {
         };
 
         runHttpAndHttps(requestHandler, () -> {
-            MyEntity result = client.doGET("/path", TypeProviders.of(MyEntity.class)).toBlocking().first();
+            MyEntity result = client.doGET("/path", Collections.emptyMap(), TypeProviders.of(MyEntity.class)).toBlocking().first();
```
[the remark this answers is not in the document]
Comment (andrew-leung) [reply titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RxNettyRestClientTest.java]: Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryConfigurationEndpointResolver.java (hunk 3 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/RegistryConfigurationEndpointResolver.java]: We discussed this offline. There is a DNS use case as well as the base class provides an async interface, so going to leave for now.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidator.java (hunk 4 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidator.java]: Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java (hunk 5 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java]: Will do

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java (hunk 8 above)
Comment (andrew-leung) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/AppNameSanitizer.java]: Good idea, will do.

File: titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RetryableRestClient.java (hunk 12 above)
Comment (andrew-leung) [reply titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RetryableRestClient.java]: Good idea, will do.

File: titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RxNettyRestClient.java (hunk 13 above)
Comment (andrew-leung) [reply titus-common/src/main/java/com/netflix/titus/common/network/client/internal/RxNettyRestClient.java]: Yup.

File: titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RetryableRestClientTest.java (hunk 15 above)
Comment (andrew-leung) [reply titus-common/src/test/java/com/netflix/titus/common/network/client/internal/RetryableRestClientTest.java]: Yeah, good idea.

File: titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/PassJobValidator.java (hunk 16 above)
Comment (andrew-leung) [reply titus-api/src/main/java/com/netflix/titus/api/jobmanager/model/job/validator/PassJobValidator.java]: Not my typo :-) but changed anyways. I suppose I did miss IntelliJ's squiggly spelling line.

File: titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobSubmitAndControlNegativeTest.java (hunk 17 above)
Comment (andrew-leung) [reply titus-server-master/src/test/java/com/netflix/titus/master/integration/v3/job/JobSubmitAndControlNegativeTest.java]: Accident. I'll remove. There shouldn't be any other changes to that file.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java (hunk 18 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java (hunk 19 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java (hunk 20 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: That is my typo :-) Will fix.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java (hunk 21 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: Yeah. No idea why I retyped the whole thing...Will fix.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java (hunk 23 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/connector/registry/DefaultDockerRegistryClient.java]: Will do.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/AggregatingValidator.java (hunk 24 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/AggregatingValidator.java]: That is my typo :-) Will fix.

File: titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java (hunk 25 above)
Comment (andrew-leung) [reply titus-server-runtime/src/main/java/com/netflix/titus/runtime/endpoint/validator/JobImageValidatorConfiguration.java]: Fixed above as part of @tbak 's comments.

**Comment by andrew-leung** [comment]
@tbak @corindwyer @gabrielhartmann , I believe I'd addressed or replied to all of the comments. Please let me know if I missed any and if any other changes are suggested. Thanks for taking a look.
